### DISCUSSION TITLE
Allow way for authors to pin a comment on top

### DISCUSSION THREAD

**Discussion by ananyaneogi** [forem/forem#15912] [In Progress]
A lot of times readers add some useful comment relating to the post and the author might want that to be highlighted and not be lost in other comments. Youtube has a feature of pinned comment, so this would be similar to that.

The UI can be similar to how pinned articles look on the profile!

Also I'm not aware if there has been similar feature requests, I didn't find any on my own so I opened this issue. If this is a duplicate or this feature is already in the works then please close this 🙂  

PS: Dev.to is awesome ❤️

**Comment by Link2Twenty** [comment]
Discussion post has been created on Forem

https://forem.dev/thomasbnt/pin-a-message-fpo

**Comment by benhalpern** [comment]
This is a great idea.

We can work on the backend functionality on this without having to  entirely figure out the design right away.

This could end up being a bit of a pain from an implementation perspective....

I'd say we could do this by adding a `pinned` boolean to comments, and then automatically bump up the `score` of the pinned comment to ensure it rises to the top (or do this otherwise, but I'd be a little concerned about having an overly complicated sorting functionality, or needing multiple queries)...

Other thoughts... Should pinning or something be a type of "reaction" which could make this a little extensible... Probably not, just thinking outloud.

**Comment by Link2Twenty** [comment]
Would pinned be a comment flair #9620?

**Comment by Taniyah-L-Jackson** [comment]
Sounds like a good idea.

The setup could be similar to StackOverflow or Medium's UI, with a unique design of course.

**Comment by nickytonline** [comment]
As discussed, I'm assigning this to you @Taniyah-L-Jackson.

**Comment by nickytonline** [comment]
@benhalpern, @rhymes and Ruby folk. Any reason we couldn't store the pinned comment ID on the article entity? Is it because of caching issues in regards to an article? To me it makes more sense there because it makes the relationship clearer, but `pinned` on a comment entity could make sense as well as @benhalpern mentioned.

I'm going to be pairing on this with @Taniyah-L-Jackson this week, so trying to get a general idea of where this functionality on the backend should live.

**Comment by rhymes** [comment]
My proposal is the following:

I’m looking at [ProfilePin](https://github.com/forem/forem/blob/master/app/models/profile_pin.rb) and I wonder if it can be recycled for this by having `profile` extended to `Article` and `pinnable` extended to `Comment`. Something like:

```ruby
class ProfilePin < ApplicationRecord
  belongs_to :pinnable, polymorphic: true
  belongs_to :profile, polymorphic: true

  validates :profile_id, presence: true
  validates :profile_type, inclusion: { in: %w[Article User] } 
  validates :pinnable_id, presence: true, uniqueness: { scope: %i[profile_id profile_type pinnable_type] }
  validates :pinnable_type, inclusion: { in: %w[Article Comment] }
  validate :only_five_pins_per_profile, on: :create
  validate :pinnable_belongs_to_profile

  # ...
end
```

So an article can have a comment pinned to it, and in theory multiple ones. This way we don't need to tamper with the score. 

There's another aspect, regardless of the implementation: what do we do with duplication? When we select the trees of comments we shouldn't remove the pinned comment from it. For example:

```markdown
- pinned comment A 
- pinned comment B
- comment 1
  - comment 2
    - comment A
    - comment 3
- comment B
```

`comment A` shouldn't be removed from its thread otherwise the flow of information fails and any comment that refers to it or any comment quoting it would lose meaning.

For this reason I think that having a different entity which lists pinned comments it's best.

I'd imagine something like (pseudo code):

`article.pinned_comments` which loads `ProfilePin.where(profile: article, pinnable_type: "Comment")` from the DB. The IDs can be cached in case the additional query is too slow but we can get to it in the PR if needed

**Comment by nickytonline** [comment]
Definitely keep e.g. `comment A` in its thread.

Also, if `comment A` has replies do we keep that whole thread under it in the pinned area as well? I would think so, but then the comment thread starting at comment A is in two places now. 🤔

**Comment by citizen428** [comment]
> Any reason we couldn't store the pinned comment ID on the article entity?

tl;dr (from a somewhat purist OO perspective): it's the comment that gets pinned, so having to modify another entity to accommodate this seems a bit off. 

More detail: We already have more than one commentable type (articles and podcast episodes) and we may conceivably add more in the future. I wouldn't want to spread that logic across all of these models or have to remember to include a concern or something like that. By keeping everything [1] in comments (or potentially with something like @rhymes suggested) we can e.g. have a service that returns the comment tree without the commentable itself knowing anything about it.

[1] "everything" here means for example: comments know how to pin/unpin themselves, comment partials add pin/unpin buttons that the author of the commentable can see, etc. 

> Also, if comment A has replies do we keep that whole thread under it in the pinned area as well?

I'd probably just add a "see in context" link that takes you to the subtree. That's quick, relatively intuitive and should be easy to implement. We can always iterate on it later. 

Update: I just noticed that Slack does something similar when you use the "Also send to #channel" option in a thread:
<img width="1046" alt="Screenshot_15_9_20_14_55" src="https://user-images.githubusercontent.com/47985/93182195-990fe480-f763-11ea-8e69-ddaba56b587f.png">

**Comment by rhymes** [comment]
@nickytonline 

> Also, if comment A has replies do we keep that whole thread under it in the pinned area as well? I would think so, but then the comment thread starting at comment A is in two places now. 🤔

I would say no, after all the goal of pinning is to let people reach useful information correlated to the post. If the author wants one of the replies to surface, they'd pin that too. Threads can be long, if I pin a popular comment with many responses, then the goal of pinning gets somewhat void as you're essentially pinning the whole comment area

**Comment by Taniyah-L-Jackson** [comment]
I do have a question about `pinnable_type` and `profile_type`. There seems to be a possibility that they could both reference `Article`. Would it be best to add a function that outputs an error if that were to happen?

**Comment by rhymes** [comment]
@Taniyah-L-Jackson makes sense, a validator that checks that they don't both point to article as you said :)  thanks!

**Comment by Taniyah-L-Jackson** [comment]
Morning, I wanted to ask about how to display the pinned comments. Instead of the comment surfacing to the top, could an auto-scroll be useful? Similar to how PDF word search works, but instead of the search bar in the image, it shows '1 out of 5 pins'.  Also, if it were possible, would this route be difficult or fairly simple? I'm wondering if it would cause the code to be clunky and 'overworded' or neat and straight to the point

<img width="567" alt="pdfExample" src="https://user-images.githubusercontent.com/52177479/94946836-51e35c80-04a2-11eb-957e-69e991f63146.png">

**Comment by rhymes** [comment]
Hi @Taniyah-L-Jackson - I think we should stick to the original idea. The goal here is to feature prominent comments right after the article. 

The problem with autoscroll is that per se it wouldn't visually help the user, especially in a long conversation, unless you highlight the comment the scroll is trying to point to. In addition, what would be the autoscroll trigger? It's not a search function, I envision it more akin the featured articles in the read next section:

![Screenshot_2020-10-03 That's so Rusty Smart pointers - DEV Community 👩‍💻👨‍💻](https://user-images.githubusercontent.com/146201/94983281-a1c53080-0541-11eb-8e02-239e2036c2dd.png)

So, for the example of pinned comments I think they should leave in a box of their own, similar to the aforamentioned "read next". I'm not entirely sure if such box should be before or after the comment form but one of our designers (cc @ludwiczakpawel @lisasy) can weigh in here or when the PR is ready as to the positioning.

![Screenshot_2020-10-03 What was your win this week - DEV Community 👩‍💻👨‍💻](https://user-images.githubusercontent.com/146201/94983323-f9639c00-0541-11eb-96bf-22a8d91fec2e.png)

Let us know if you have further questions

**Comment by Taniyah-L-Jackson** [comment]
@rhymes For the `articles` and `comments` schema, will an FK boolean table be helpful in checking which elements are pinned? (Possible set-up: `  create_table "isPinned", do |t|
    t.boolean "pinned", default: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end`)
Then, agreeing with the pseudo-code you wrote earlier, the articles and comments can go through validations to separate the pinned comments from the articles. As far as displaying them, I'm wondering if sorting the data into two new tables or creating joining tables will work.

**Comment by rhymes** [comment]
Hi @Taniyah-L-Jackson, you don't really need a new table. `ProfilePin` indicates that the item has been pinned. The combination of `profile_id` and `profile_type` tells you if it was pinned on an article or on a user profile and the combination of `pinnable_id` and `pinnable_type` tells you if the pinned item is an article or a comment. In this case you need `profile_type` to be `Article` and `pinnable_type` to be `Comment` with their relative IDs.

The mere existence of a row in that table answer the questions "is it pinned?".

I believe this is the simplest solution with the minimum amount of restructuring involved. 

To retrieve the list of "pinned comments" you need to add a relation to the `Article` model that points to `ProfilePin` so you can do something like `article.profile_pins` where the query will be something like (pseudo code): `select all rows from profile_pins where profile_id is this article, profile_type is 'Article' and pinnable_type is 'Comment'`.

This query will give you a list of comments ids that you can then load from the DB to display the comments themselves. Probably sorted by decreasing `score`

Let me know if anything is unclear

**Comment by Taniyah-L-Jackson** [comment]
Afternoon @rhymes! Thank you for the earlier clarification, my understanding was still a bit hazy. Would adding the line `through: :profile-pin` to `has_many :profile_pins, as: :pinnable, inverse_of: :pinnable, dependent: :destroy` be considered as a relation between the two? Or does this need to be fleshed out more?

**Comment by rhymes** [comment]
Hi @Taniyah-L-Jackson, I'm not sure you need to add any clauses to the relation. If we could reason around a PR it'd probably be easier for us to help get you unstuck on the implementation. It's a bit harder otherwise 😊

**Comment by techieeliot** [comment]
Hey, @Taniyah-L-Jackson I'm available to help pair with you (and @nickytonline maybe) and let's get this PR wrapped up for hacktoberfest for you. Hope you're doing well.

**Comment by abhijit-hota** [comment]
@rhymes 
> I think they should leave in a box of their own

I feel that a pinned comment should be just that: a comment which is pinned to the top along with a flair that says **"Pinned by Author"** or anything along those lines. As @benhalpern  said, 

> we could do this by adding a `pinned` boolean to comments

...and displaying that particular comment (thread) in the beginning of the **Discussion** section. Simplest example is probably YouTube.

P.S. I came here to file a feature request for this but ended up reading this thread. Just my two cents! 😄

**Comment by cmgorton** [comment]
Hey @Taniyah-L-Jackson  
We are working on changing how we approve feature requests and when the Forem team or contributors will take them on.
I am going to go a head and unassign you to this feature for now while we work through our [new RFC process](https://forem.dev/foremteam/internal-rfc-process-and-forem-dev-discussions-3gl4). Thank you fo your help here!

In the meantime if you see any bugs you want to help squash let me know and we can assign you to them!

**Comment by Taniyah-L-Jackson** [comment]
Understood, will do 😊

**Comment by yugal95** [comment]
Is this issue still active or in consideration?

**Comment by cmgorton** [comment]
Thanks for asking! I believe it is still in consideration. 
We recently released a feature for admins to pin an article and this is a similar feature but for authors of posts. I'll ping @forem/product here to hear their thoughts.

**Comment by ellativity** [comment]
Keeping this one alive! I wish I'd been able to pin [this comment](https://forem.dev/lee/comment/4in) by @leewynne on forem.dev

**Comment by michael-tharrington** [comment]
No doubt! This is seriously a great request and jives well with our ethos around giving content creators more control over how they present their content.🙌 

I also think this would work really well for Q/A software. If I'm not mistaken, I feel like we've talked internally about making a feature that would allow someone to mark something as a "best answer"... in a more generic sense, I would think the functionality could work the same here. Maybe the author could choose to pin a comment with or without the messaging of "best answer".

Side note: would be fun to have a badge for getting your comment marked as a best answer by someone else.

**Reply by ildi-x** [reply]
I love the idea of earning badges for writing valuable comments. What sort of threshold would trigger the badge? Would it be something like once a comment receives 10 or more likes? I was just looking over the [badge docs](https://admin.forem.com/docs/forem-basics/badges) and it seems like Dev already has something called the **beloved comment**:  [this badge](https://dev.to/badge/beloved-comment) is awarded automatically for any comment that reaches a certain threshold of positive reactions. 

This does make me wonder about who should be able to pin a comment within a post, should it be the author of that particular post or should it be signaling from other community members, like positive reactions? 

@michael-tharrington adding to your side note, it might be cool if users could send/reward badges to each other when they find a particular comment valuable. It can be an option in the comment footer where we currently have like + reply buttons or it can be included within the comment dropdown trigger. To make this work, when a user first joins a Forem community, they are given a certain amount of "thank you" badges which they can send to others when they find a comment to be valuable. Who wouldn't want to be a collector of the most prestigious Forem badges? 😍

**Reply by michael-tharrington** [reply]
Lots of good thoughts, Ildi! I was thinking that it would make most sense for the author to be able to pin the comment, but I also totally realize that this would make attaching a badge to that action pretty unique as it would allow authors to award a badge, which could be game-able.

All this said, I do really love your idea for a thank you badge!

I've had a similar thought around the unicorn reaction previously. I kinda wish that there were a limited number of unicorns and when you received one, you could give it away to someone else but would lose it in the process. Shoot maybe there could even be badges for giving and receiving a unicorn though!

Anyway, lots to chew on here. 🤔 😁

**Reply by ildi-x** [reply]
There are certainly some interesting ideas here!  

Now that I've been participating within Forem/Dev for a while, I have not seen the value of the unicorn reaction. I kind of feel guilty to only click the like reaction when I enjoy a post, so I end up also clicking on the unicorn as well. 

I really like the idea of having a limited number of badges/unicorns within a community that continue to circulate and trade hands as a reward for users being helpful. 

I think of Forem badges similarly to how roles work within Discord. If i'm a new user of a Forem community, it might be valuable to get a reading on who are the most distinguished members. At the moment, we do give admins the option to hand-pick/suggest a few users during the onboarding process, but there is no quick way of finding out who the most active/engaged members are in the community, or who has been providing a lot of value. I guess in a way, this type of info is derived by being an active member yourself as you spend more time within a community. Maybe we don't want to rank users by how many posts or comments they are making, but I think it might be interesting to filter users by the badges that they hold. This in turn would also bring more attention to badges. Also right now it's not very clear who the admins + mods of a community are. Maybe a user figures it out during the onboarding, welcome thread, or inside the /about page, but I think this info needs to be easier to get to. 

I do like that Forem allows users to search for people (other members), but this feature is kind of hidden behind the search bar at the moment.

**Comment by amywtlin** [comment]
Just so people here are aware, we do have something brewing for this: #16306
